From 444970822fc438f4a86f8d331f4eba2ef82fa3ca Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Thu, 8 Dec 2022 10:26:26 +0800
Subject: [PATCH 79/98] HACK: vnc: Make pam optional

Allow building VNC backend without PAM, but force TLS disable for security.
This enables VNC usage in environments where PAM is unavailable.

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 frontend/main.c             | 5 +++++
 libweston/backend-vnc/vnc.c | 5 +++++
 libweston/meson.build       | 7 ++++---
 3 files changed, 14 insertions(+), 3 deletions(-)

diff --git a/frontend/main.c b/frontend/main.c
index d3001b5e..b2f060a7 100644
--- a/frontend/main.c
+++ b/frontend/main.c
@@ -4733,6 +4733,11 @@ load_vnc_backend(struct weston_compositor *c,
 		{ WESTON_OPTION_BOOLEAN, "disable-transport-layer-security", 0, &config.disable_tls },
 	};
 
+	/* Force disable TLS when built without PAM for security */
+#ifndef HAVE_PAM
+	config.disable_tls = true;
+#endif
+
 	parse_options(vnc_options, ARRAY_LENGTH(vnc_options), argc, argv);
 
 	config.renderer = renderer;
diff --git a/libweston/backend-vnc/vnc.c b/libweston/backend-vnc/vnc.c
index ae689d85..34fd8a82 100644
--- a/libweston/backend-vnc/vnc.c
+++ b/libweston/backend-vnc/vnc.c
@@ -1328,12 +1328,17 @@ vnc_backend_create(struct weston_compositor *compositor,
 
 		weston_log("TLS support activated\n");
 	} else {
+#ifdef HAVE_PAM
 		ret = nvnc_enable_auth(backend->server, NVNC_AUTH_REQUIRE_AUTH,
 				       vnc_handle_auth, NULL);
 		if (ret) {
 			weston_log("Failed to enable authentication\n");
 			goto err_nvnc;
 		}
+#else
+		weston_log(
+			"warning: VNC enabled without auth!\n");
+#endif
 
 		weston_log(
 			"warning: VNC enabled without Transport Layer "
diff --git a/libweston/meson.build b/libweston/meson.build
index a4124358..44e96761 100644
--- a/libweston/meson.build
+++ b/libweston/meson.build
@@ -109,10 +109,11 @@ if get_option('backend-vnc')
 		dep_pam = cc.find_library('pam', required: false)
 	endif
 	if not dep_pam.found()
-		error('VNC backend requires libpam which was not found. Or, you can use \'-Dbackend-vnc=false\'.')
+		warning('VNC backend suggests libpam which was not found. Or, you can use \'-Dbackend-vnc=false\'.')
+	else
+		config_h.set('HAVE_PAM', '1')
+		deps_libweston += dep_pam
 	endif
-	config_h.set('HAVE_PAM', '1')
-	deps_libweston += dep_pam
 endif
 
 if get_option('perfetto')
-- 
2.39.5

